libraries/chain/peer_keys_db.cpp

```diff
@@ -0,0 +1,74 @@
+#include <eosio/chain/peer_keys_db.hpp>
+#include <eosio/chain/contract_table_objects.hpp>
+
+namespace eosio::chain {
+
+peer_keys_db_t::peer_keys_db_t()
+   : _peer_key_map(boost::make_shared<peer_key_map_t>()) {
+}
+
+std::optional<uint64_t> peer_keys_db_t::_get_version(const chainbase::database& db) {
+   const auto* t_id =
+      db.find<table_id_object, by_code_scope_table>(boost::make_tuple("eosio"_n, "eosio"_n, "peerkeysver"_n));
+   if (t_id == nullptr)
+      return std::optional<uint64_t>{};
+
+   const auto& idx = db.get_index<chain::key_value_index, chain::by_scope_primary>();
+
+   auto itr = idx.lower_bound( boost::make_tuple(t_id->id, 0));
+   if (itr == idx.end() || itr->t_id != t_id->id || itr->primary_key != 0)
+      return std::optional<uint64_t>{};
+   
+   const key_value_object& obj = *itr;
+   fc::datastream<const char*> ds(obj.value.data(), obj.value.size());
+   uint64_t res = 0;
+   fc::raw::unpack(ds, res);
+   return std::optional<uint64_t>(res);
+}
+
+size_t peer_keys_db_t::update_peer_keys(const controller& chain) {
+   size_t num_updated = 0;
+   try {
+      const auto& db = chain.db();
+      auto cb_version = _get_version(db);
```

Comment:
I am not really seeing any benefit to using chainbase's revision over just simply the applied block num? That's really what the contract is tracking -- the last block num the row was modified at. (`version` seems a misnomer in that aspect) Relying on chainbase revision == head block num might be an implementation detail.

Reply:
This tracks the version of the `peerkeys` table, and this version is incremented every time the `peerkeys` table if modified.
What I meant with `cb_version` is `the version of peerkeys in chainbase`, vs the version of `peerkeys` we currently have in the `unordered_flat_map`.

Reply:
But afaict you're just using chainbase's revision as a proxy for current block num. I don't believe that's a pattern used elsewhere -- normally you'd look at forkdb or the block passed in accepted_block.

Reply:
> just using chainbase's revision as a proxy for current block num

Where do I do that? `peer_keys_db_t::_get_version()` returns the version stored in the `peerkeysver` table.

Reply:
whoops you are right I misread the earlier code